Report error for invalid hexadecimal digits: 0x12abq

quick-lint-js reports an error if a decimal or binary number literal contains bogus digits:

```javascript
// hello.js:2:5: error: unexpected characters in number literal
1234abcd
```
Give hexadecimal literals the same error-reporting treatment. See parse_binary_number in lex.cpp, and tests in test-lex.cpp.

--- a/src/include/quick-lint-js/lex.h
+++ b/src/include/quick-lint-js/lex.h
@@ -244,6 +244,7 @@ class lexer {
 
   void parse_binary_number();
   void parse_hexadecimal_number();
+  const char8* check_garbage_in_number_literal(const char8* input);
   void parse_number();
   static const char8* parse_decimal_digits(const char8*) noexcept;
 

--- a/src/lex.cpp
+++ b/src/lex.cpp
@@ -615,9 +615,13 @@ const char8* lexer::end_of_previous_token() const noexcept {
 
 void lexer::parse_hexadecimal_number() {
   QLJS_ASSERT(this->is_hex_digit(this->input_[0]) || this->input_[0] == '.');
-  while (this->is_hex_digit(this->input_[0])) {
-    this->input_ += 1;
+  const char8* input = this->input_;
+
+  while (this->is_hex_digit(*input)) {
+    input += 1;
   }
+
+  this->input_ = check_garbage_in_number_literal(input);
 }
 
 void lexer::parse_binary_number() {
@@ -628,13 +632,17 @@ void lexer::parse_binary_number() {
     input += 1;
   }
 
+  this->input_ = check_garbage_in_number_literal(input);
+}
+
+const char8* lexer::check_garbage_in_number_literal(const char8* input) {
   const char8* garbage_begin = input;
   for (;;) {
     switch (*input) {
-    QLJS_CASE_DECIMAL_DIGIT:
-    QLJS_CASE_IDENTIFIER_START:
-      input += 1;
-      break;
+      QLJS_CASE_DECIMAL_DIGIT:
+      QLJS_CASE_IDENTIFIER_START:
+        input += 1;
+        break;
       default:
         goto done_parsing_garbage;
     }
@@ -647,7 +655,7 @@ done_parsing_garbage:
     input = garbage_end;
   }
 
-  this->input_ = input;
+  return input;
 }
 
 void lexer::parse_number() {

--- a/test/test-lex.cpp
+++ b/test/test-lex.cpp
@@ -191,6 +191,20 @@ TEST(test_lex, lex_number_with_trailing_garbage) {
                               error_unexpected_characters_in_number, characters,
                               offsets_matcher(&input, 3, 8))));
   }
+
+
+  {
+    error_collector v;
+    padded_string input(u8"0xabjjw");
+    lexer l(&input, &v);
+    EXPECT_EQ(l.peek().type, token_type::number);
+    l.skip();
+    EXPECT_EQ(l.peek().type, token_type::end_of_file);
+
+    EXPECT_THAT(v.errors, ElementsAre(ERROR_TYPE_FIELD(
+        error_unexpected_characters_in_number, characters,
+        offsets_matcher(&input, 4, 7))));
+  }
 }
 
 TEST(test_lex, lex_invalid_big_int_number) {

